### PR TITLE
Feature - UI Test with KIF

### PR DESCRIPTION
**Pull request by nmint8m**
## Why

Demo UI testing with KIF framework.

## Insight

Included in this PR:

- Integrate KIF library via Pods
- Add UI testing utility classes and functions
- Define identifiers and implement identifiers for UI components
- Define test flows
- Implement test cases and test suites
- Adjust test targets and schemes

## Proof

image

### CODE REVIEW

File: DemoViper/Sources/Extensions/UIBarButtonItem+Identifier.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//
+//  Created by Nguyen M. Tam on 11/01/2021.
+//  Copyright © 2021 Nimble Co. Ltd. All rights reserved.
+//
+
+import UIKit
+
+extension UIBarButtonItem {
+
+    func setIdentifier(_ identifier: String) {
```
Comment (markgravity) [inline DemoViper/Sources/Extensions/UIBarButtonItem+Identifier.swift:13]: What do you think using a protocol for this? Example
```swift
protocol AccessibilitySettable {
    func setIdentifier(_ identifier: String)
}

extension UIView: AccessibilitySettable {
    func setIdentifier(_ identifier: String) {
        accessibilityIdentifier = identifier
        accessibilityLabel = identifier
    }
}

extension UIBarButtonItem: AccessibilitySettable {
    func setIdentifier(_ identifier: String) {
        accessibilityIdentifier = identifier
        accessibilityLabel = identifier
    }
}

// then our UIViewController extension
extension UIViewController {
    
    func setIdentifiersForViews(with info: [String: AccessibilitySettable?]) {
        info.forEach { (identifier, view) in
            view?.setIdentifier(identifier)
        }
    }
}
```

**Review by markgravity — APPROVED** [review]
LGTM

File: DemoViperUITestsKIF/Specs/Flows/GeneralUIFlows.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                case .failure:
+                    print("Backing to root view has been failed. Error: \(error?.localizedDescription ?? "")")
+                case .wait:
+                    print("Backing to root view has been waited.")
+                default:
+                    break
+                }
+            }, timeout: TimeInterval.long)
+        } catch {
+
```
Comment (markgravity) [inline DemoViperUITestsKIF/Specs/Flows/GeneralUIFlows.swift:44]: Can we print out the error here? For the unlucky case 😄

File: DemoViperUITestsKIF/Specs/ListReposUISpecs.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            it("open the list repo screen successfully") {
+               let view = tester.waitForView(withAccessibilityLabel: ListRepos.view.identifier)
+                expect(view).notTo(beNil())
+            }
+
+            context("no repo is added to bookmark") {
+
+                it("verify initial data of a repo and open the detail repo screen") {
+                    let repoCellItem = tester.tapRepoCell(at: 0)
+                    expect(repoCellItem?.repoFullName) != ""
```
Comment (markgravity) [inline DemoViperUITestsKIF/Specs/ListReposUISpecs.swift:37]: What do you think about adding `isEmpty` property into `Repo` model? 
I saw we have a duplicate code at line 52

File: DemoViperUITestsKIF/Specs/Flows/ListRepoUIFlows.swift [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        return RepoCellItem(
+            repoFullName: cell.textLabel?.text ?? "",
+            isBookMark: cell.accessoryType == .checkmark
+        )
+    }
+}
+
+struct RepoCellItem {
+    let repoFullName: String
+    let isBookMark: Bool
```
Comment (markgravity) [inline DemoViperUITestsKIF/Specs/Flows/ListRepoUIFlows.swift:40]: What do you think about renaming `isBookMark` to `isBookmarked`?

File: DemoViper/Sources/Extensions/UIBarButtonItem+Identifier.swift (hunk 1 above)
Comment (nmint8m) [reply DemoViper/Sources/Extensions/UIBarButtonItem+Identifier.swift:13]: Thanks for your advise. Let me give a try. 👍👍👍

**Review by markgravity — APPROVED** [review]
LGTM

File: DemoViperUITestsKIF/Helper/KIFTestActor+Setup.swift [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+//
+//  Created by Nguyen M. Tam on 12/01/2021.
+//  Copyright © 2021 Nimble Co. Ltd. All rights reserved.
+//
+
+import KIF
+
+@testable import DemoViper
+
+func setupTestEnvironment() {
```
Comment (blyscuit) [inline DemoViperUITestsKIF/Helper/KIFTestActor+Setup.swift:13]: I see setUp and setup being interchangeably used here, very minor detail.

**Review by blyscuit — APPROVED** [review]

File: DemoViper/Sources/Modules/ListRepos/View/ReposViewController.swift [hunk 6]
```diff
@@ -80,6 +80,7 @@ extension ReposViewController {
         setupNavigationBar()
         setupTableView()
         setupEmptyView()
```
Comment (suho) [inline DemoViper/Sources/Modules/ListRepos/View/ReposViewController.swift:82]: Please update `setUp` for these functions

Comment (nmint8m) [reply DemoViper/Sources/Modules/ListRepos/View/ReposViewController.swift:82]: I renamed those functions in PR #15. Please check it out.

**Review by suho — APPROVED** [review]

**Review by phongvhd93 — APPROVED** [review]

**Pull request merged**